File: guides/micronaut-hibernate-reactive/java/src/main/java/example/micronaut/SortingAndOrderArguments.java
package example.micronaut;

import io.micronaut.core.annotation.Nullable;
import io.micronaut.serde.annotation.Serdeable;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.PositiveOrZero;

@Serdeable
public record SortingAndOrderArguments(
        @Nullable @PositiveOrZero Integer offset,
        @Nullable @Positive Integer max,
        @Nullable @Pattern(regexp = "id|name") String sort,
        @Nullable @Pattern(regexp = "asc|ASC|desc|DESC") String order
) {
}

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/SortingAndOrderArguments.kt
package example.micronaut

import io.micronaut.serde.annotation.Serdeable
import jakarta.validation.constraints.Pattern
import jakarta.validation.constraints.Positive
import jakarta.validation.constraints.PositiveOrZero

@Serdeable
data class SortingAndOrderArguments(
    @field:PositiveOrZero var offset:  Int? = null, // <1>
    @field:Positive var max: Int? = null, // <1>
    @field:Pattern(regexp = "id|name") var sort: String? = null, // <1>
    @field:Pattern(regexp = "asc|ASC|desc|DESC") var order: String? = null // <1>
)